src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,315 @@
+          options = Aws::SDKOptions();
+          Aws::InitAPI(options);
+    }
+    // dtor wait for thread to stop
+    // then connection are cleaned-up
+    ~Manager() {
+      stopped = true;
+      runner.join();
+      messageQueue.consume_all(delete_message);
+      Aws::ShutdownAPI(options);
+    }
+
+
+    AwsClient connect(const std::string &accessKey,
+                      const std::string &accessSecret,
+                      const std::string &arn,
+                      const std::string &endpoint,
+                      bool verifySSL,
+                      bool useSSL) {
+      if (stopped) {
+        return NO_CLIENT;
+      }
+      const client_id_t id(accessKey, accessSecret, arn, useSSL);
+      std::lock_guard lock(clientsLock);
+      const auto it = clientList.find(id);
+      if (it != clientList.end()) {
+        return it->second;
+      }
+
+      // connection not found, creating a new one
+      Aws::Client::ClientConfiguration configuration;
+      if (useSSL) {
+        configuration.scheme = Aws::Http::Scheme::HTTPS;
+      } else {
+        configuration.scheme = Aws::Http::Scheme::HTTP;
+      }
+      configuration.verifySSL = verifySSL;
+      Aws::Utils::ARN arnAWS((Aws::String(arn)));
+      configuration.region = Aws::String(arnAWS.GetRegion());
+      configuration.maxConnections = MAX_CONNECTIONS;
+      configuration.requestTimeoutMs = REQUEST_TIMEOUT_MS;
+      configuration.connectTimeoutMs = CONNECT_TIMEOUT_MS;
+      std::string host, user, password;
+      rgw::parse_url_authority(endpoint, host, user, password);
+      configuration.endpointOverride = Aws::String(host);
+      Aws::Auth::AWSCredentials credentials;
+      credentials.SetAWSAccessKeyId(Aws::String(accessKey));
+      credentials.SetAWSSecretKey(Aws::String(accessSecret));
+      AwsClient client;
+      if (arnAWS.GetService() == "lambda") {
+        client = new Aws::Lambda::LambdaClient(credentials, configuration);
+      } else if (arnAWS.GetService() == "sns") {
+        client = new Aws::SNS::SNSClient(credentials, configuration);
+      } else {
+        return NO_CLIENT;
+      }
+      return clientList.emplace(id, client).first->second;
+    }
+
+    int publish(AwsClient client,
+                const std::string &payload,
+                const std::string &resourceARN,
+                reply_callback_t cb) {
+      if (stopped) {
+        return STATUS_MANAGER_STOPPED;
+      }
+      if (client == NO_CLIENT) {
+        return STATUS_CLIENT_CLOSED;
+      }
+      if (messageQueue.push(new message_t(client, resourceARN, payload, cb))) {
+        return STATUS_OK;
+      }
+
+
+      return STATUS_FULL;
+    }
+
+    void stop() {
+      stopped = true;
+    }
+
+  };
+
+  static Manager *manager = nullptr;
+  static const unsigned IDLE_TIME_MS = 100;
+  static const size_t MAX_QUEUE_DEFAULT = 8192;
```

Comment:
what happens when the queue is full?

Reply:
> what happens when the queue is full?

same as in the cases of kafka and amqp. the notification fails.
with non-persistent notifications this is just logged as an error, with persistent notifications this will be retried later.
(when the persistent retry queue is exhausted this will fail the operation that triggers the notification with 503)